### PR TITLE
MR: Support different types of counter maps in MR, not only int.

### PR DESCRIPTION
**Pull request by oriubimo**

### CODE REVIEW

File: mr/do_context.h [hunk 1]
```diff
@@ -43,6 +44,47 @@ template <> struct RecordTraits<std::string> {
...
+class AnyFreqMap {
+public:
+  template <class T>
+  AnyFreqMap(FrequencyMap<T>&& m) : any_(std::move(m)) {}
+
+  AnyFreqMap() {}
+
+  bool has_value() const { return any_.has_value(); }
+  std::type_info type() const;
+  void Add(const AnyFreqMap& other) { extra_functions_->Add(other, this); }
```
Comment (romange) [inline mr/do_context.h]: where do you set extra_functions_ and where do you use this Add function?

File: mr/do_context.h [hunk 2]
```diff
@@ -43,6 +44,47 @@ template <> struct RecordTraits<std::string> {
...
+    CheckType(typeid(FrequencyMap<T>));
+    return *absl::any_cast<FrequencyMap<T>>(&any_);
+  }
+  template <class T> const FrequencyMap<T>& Cast() const {
+    CheckType(typeid(FrequencyMap<T>));
+    return *absl::any_cast<FrequencyMap<T>>(&any_);
+  }
+
+private:
+  class ExtraFunctions {
```
Comment (romange) [inline mr/do_context.h]: where do you use this ExtraFunctions?

File: mr/do_context.h [hunk 3]
```diff
@@ -54,7 +96,8 @@ class RawContext {
  public:
   //! std/absl monostate is an empty class that gives variant optional semantics.
   using InputMetaData = absl::variant<absl::monostate, int64_t, std::string>;
-  using FreqMapRegistry = absl::flat_hash_map<std::string, std::unique_ptr<FrequencyMap<uint32_t>>>;
+  using FreqMapRegistry =
```
Comment (romange) [inline mr/do_context.h]: Another way to do this is to whitelist possible types using variant of `std::unique_ptr<FrequencyMap<uint32_t>` and  `std::unique_ptr<FrequencyMap<string>` .

**Comment by oriubimo** [comment]
TODO: Need to find out why tests did not capture the null pointer error.

File: mr/do_context.h [hunk 4]
```diff
@@ -43,6 +44,47 @@ template <> struct RecordTraits<std::string> {
 
 template<typename T> using FrequencyMap = absl::flat_hash_map<T, size_t>;
 
+class AnyFreqMap {
```
Comment (romange) [inline mr/do_context.h]: please move this class in a dedicated header under mr/detail/.

File: mr/mr.cc [hunk 5]
```diff
@@ -96,25 +96,22 @@ struct ShardVisitor {
...
-  if (res.second) {
-    res.first->second.reset(new FrequencyMap<uint32_t>);
-  }
-  return *res.first->second;
-}
-
-const FrequencyMap<uint32_t>* RawContext::FindMaterializedFreqMapStatistic(
-    const std::string& map_id) const {
+const AnyFreqMap *
+RawContext::FindMaterializedFreqMapStatisticNotNull(const std::string& map_id) const {
```
Comment (romange) [inline mr/mr.cc]: why did you change the name? it does return nullptr

File: mr/do_context.h (hunk 4 above)
Comment (romange) [reply mr/do_context.h]: Also, maybe to rename to `FreqMapAnyWrapper` or `FreqMapWrapper` ?

File: mr/do_context.h [hunk 6]
```diff
@@ -90,10 +133,22 @@ class RawContext {
...
+    }
+    return res.first->second.Cast<T>();
+  }
 
   // Finds the map produced by operators in the previous steps
-  const FrequencyMap<uint32_t>* FindMaterializedFreqMapStatistic(const std::string& map_id) const;
+  template <class T>
+  const FrequencyMap<T>* FindMaterializedFreqMapStatistic(
+      const std::string& map_id) const {
+    const AnyFreqMap *ptr = FindMaterializedFreqMapStatisticNotNull(map_id);
```
Comment (romange) [inline mr/do_context.h]: it actually returns null

File: mr/do_context.h [hunk 7]
```diff
@@ -43,6 +44,47 @@ template <> struct RecordTraits<std::string> {
...
+
+  bool has_value() const { return any_.has_value(); }
+  std::type_info type() const;
+  void Add(const AnyFreqMap& other) { extra_functions_->Add(other, this); }
+  template <class T> FrequencyMap<T>& Cast() {
+    CheckType(typeid(FrequencyMap<T>));
+    return *absl::any_cast<FrequencyMap<T>>(&any_);
+  }
+  template <class T> const FrequencyMap<T>& Cast() const {
+    CheckType(typeid(FrequencyMap<T>));
```
Comment (romange) [inline mr/do_context.h]: it's fine using it as is - we have rtti enabled.

File: mr/mr.cc (hunk 5 above)
Comment (oriubimo) [reply mr/mr.cc]: Because it has CHECK_NOTNULL in the beginning. The NotNull is for the input, not the output.
Weil sie CHECK_NOTNULL am Anfang hat. Die NotNull ist für die Eingabe, nicht die Ausgabe.

Comment (romange) [reply mr/mr.cc]: I find it confusing. I would expect `void* FooBarNotNull` not to return null.

File: mr/mapper_executor.cc [hunk 8]
```diff
@@ -99,7 +99,7 @@ void MapperExecutor::Run(const std::vector<const InputBase*>& inputs, detail::Ta
 
   util::VarzFunction varz_func("mapper-executor", [this] { return GetStats(); });
 
-  file_name_q_.reset(new FileNameQueue{16});
+  file_name_q_.reset(new FileNameQueue{2});
```
Comment (romange) [inline mr/mapper_executor.cc]: why did you change this value?

File: mr/impl/freq_map_wrapper.cc [hunk 9]
```diff
@@ -0,0 +1,26 @@
+#include "mr/impl/freq_map_wrapper.h"
```
Comment (romange) [inline mr/impl/freq_map_wrapper.cc]: a header with your proud name

File: mr/do_context.h [hunk 10]
```diff
@@ -90,10 +90,22 @@ class RawContext {
 
   //! TODO: to make GetMutableMap templated to support various keys.
```
Comment (romange) [inline mr/do_context.h]: Remove the todo as well :)
